### ISSUE TITLE
Project Feedback - new lessons

### ISSUE THREAD

**Issue by glomstea** [glomstea/MI-449-SS17-741-js-objects-T4W3KQ#1]
## Build an app that can tell jokes, learn new jokes, and most importantly: forget bad jokes 😅

@stuartpearman Can you take a look at this? It's [hosted here](https://glomstea-kaprwtcydf6wbt4w3kq.surge.sh/) and meets the following criteria:

- [x] Your app builds on [these files](https://gist.github.com/chrisvfritz/3bd1b8d8e5b0d1ea2278ffbc561de047) (you can make changes to any of them)
- [x] When the `requestedJokeInput` has a value matching the key of an existing joke, the `jokeBox` HTML is updated [in this format](https://gist.github.com/chrisvfritz/4def3c246dfb931ecb57384a5639af99)
- [x] When the `requestedJokeInput` does _not_ match an existing joke, the `jokeBox` content is updated to `No matching joke found.`
- [x] When the `Remember this joke!` button is clicked, a new joke is saved (or overwritten) in the `jokes` object, using the data in the associated `input`/`textarea` elements
- [x] When the `Forget about it!` button is clicked, the joke with that key (if any) is deleted from the `jokes` object
- [x] Whenever the `jokes` object changes, the new version is saved in `localStorage` and used whenever the user visits the page again
- [x] It's impossible for an error to ever occur in the JavaScript console

Can you open up the next three lessons for me? I would like to start them asap, and focus on this revision later.

Thanks

**Comment by KatieMFritz** [comment]
Hi Grant, I'm getting this started for Stuart. Your Joke-A-Tron works pretty well! 🤖 One thing I'm not seeing work:

> Whenever the jokes object changes, the new version is saved in `localStorage` and used whenever the user visits the page again

Refer back to the `localStorage` lesson if you need a refresher, and as always, Slack us if you need help! We're all a bit busy tonight, so you could also try posting in the #mi449 channel and a classmate might also be able to help. 😁

**Comment by glomstea** [comment]
@KatieMFritz I am stuck on that part at the moment but I made progress on the future lessons. Would you mind opening them up for me?

**Comment by KatieMFritz** [comment]
We can't do that, sorry. Do you want help with local storage?

**Comment by glomstea** [comment]
@KatieMFritz Yes please. I am retrieving from localstorage fine, but the app doesn't seem to want to write to it. I just updated my code, and have been using window.localStorage.setItem('jokes', jokes) in both my add and remove function where the joke object is changed but it does not work. I have also tried putting this line in the updatePage function, and that doesn't work either. I am not sure if I am not using the right command, or if it is at the wrong place, but I have pasted that line in just about every spot in my code, but it still wont write.

**Comment by egillespie** [comment]
Hi @glomstea, the answers you're looking for are on the last page of [the lesson](https://msu.lansing.codes/courses/MI-449-SS17-741/lessons/js-objects/6). `localStorage` stores all information as strings, but the object you need to store in this lesson is an object.

Read it over thoroughly, do your best to apply the material to your project, and comment back when you're done or get stuck! ☘️

**Comment by glomstea** [comment]
After a lot of clunking around, everything seems to work. @KatieMFritz @egillespie

**Comment by egillespie** [comment]
Great, I can see everything's functional now. 👍

I also like the text that shows whether a joke was added or updated! I do have one recommendation to improve upon that portion of your code.

### Check for existing joke without loop

Instead of using a `for` loop to find an existing joke, you can instead use the key and an `if` to test whether the joke key matches a joke. Here's how `updateDisplayedJoke` would look using this approach:

```js
var updateDisplayedJoke = function () {
  var requestedJokeKey = requestedJokeInput.value
  var existingJoke = jokes[requestedJokeKey]
  if (existingJoke) { // use truthiness to determine if joke exists
    jokeBox.textContent = existingJoke.setup + ' ' + existingJoke.punchline
  } else {
    jokeBox.textContent = noJokeFound
  }
}
```

Also notice how I used `existingJoke.setup` and `existingJoke.punchline` instead of `existingJoke['setup']` and `existingJoke['punchline']`. They both do the same thing, but using `.` is faster to type, easier to read, and only works when you know the exact keys you need to look up.

In addition to `updateDisplayedJoke`, can you also update `addJoke` and `removeJoke` to use this way of determining if a joke exists?

### Put setup and punchline in their own paragraphs

Could you also update `updateDisplayedJoke` to put setup and punchline in their own paragraphs to satisfy criteria 2? Right now they are separated by a space and they run together.

---
After you’ve made your changes and pushed them to GitHub and your hosted site, give it a once-over to make sure it looks right, then comment back here and I’ll take another look.

Thanks! 🦄

**Comment by glomstea** [comment]
@egillespie I just fixed the punchline and setup criteria which my code did not meet, but changing both your other suggestions broke my app. Using truthiness broke my joke menu, and getting rid of my brackets messed up my variables. Since it is working, and still meets all the requirements, can I move onto the next lesson? I have spent far too much time on it so far.

**Comment by egillespie** [comment]
Hi @glomstea, I created a [pull request](https://github.com/glomstea/MI-449-SS17-741-js-objects-T4W3KQ/pull/2) that demonstrates how to check for the existence of a joke without a loop and how to access fields without brackets.

Since these are core concepts of this lesson, I want to make sure you can apply them before moving on to another lesson. Objects, in particular, are a fundamental concept worth understanding in JavaScript and will be used in many of the subsequent JavaScript lessons.

Go ahead and accept the pull request and see what I changed to make it work. I believe the issues you ran into were caused by typos in my example. I didn't remove the `break` statement, and I referenced the wrong object in one spot. I've updated my comment to be more correct.

### Remove loop from `removeJoke`

I did not update the `removeJoke` function. I'd still like you to correct that one so I can see that you understand the changes involved.

### Put setup and punchline in their own paragraphs

You added newlines between your setup and punchline. We would like to see paragraphs and provided a [template](https://gist.github.com/chrisvfritz/4def3c246dfb931ecb57384a5639af99) for you to use. Newlines are not an appropriate way to separate this content, as explained on the first page of the _[Present a wider variety of content in your HTML](https://msu.lansing.codes/courses/MI-449-SS17-741/lessons/html-more-kinds-of-content/1)_ lesson.

---

Give these changes another shot. Please reach out to me on Slack if you get stuck again. Comment here and ping me on Slack when you're ready for me to review. I'll jump on them ASAP.

Thanks! ⛵️

**Comment by glomstea** [comment]
Finally working. Should check out. @egillespie @KatieMFritz

**Comment by egillespie** [comment]
Beautiful! Thanks for putting the paragraph tags in your HTML file too! :shipit:
